File: src/main/java/org/hisp/dhis/lib/expression/math/GS1Formatter.java
package org.hisp.dhis.lib.expression.math;

import java.util.HashMap;
import java.util.Map;

import static org.hisp.dhis.lib.expression.math.GS1Elements.BEST_BEFORE_DATE;
import static org.hisp.dhis.lib.expression.math.GS1Elements.CONTENT;
import static org.hisp.dhis.lib.expression.math.GS1Elements.DUE_DATE;
import static org.hisp.dhis.lib.expression.math.GS1Elements.EXP_DATE;
import static org.hisp.dhis.lib.expression.math.GS1Elements.GTIN;
import static org.hisp.dhis.lib.expression.math.GS1Elements.PACK_DATE;
import static org.hisp.dhis.lib.expression.math.GS1Elements.PROD_DATE;
import static org.hisp.dhis.lib.expression.math.GS1Elements.SELL_BY;
import static org.hisp.dhis.lib.expression.math.GS1Elements.SSCC;
import static org.hisp.dhis.lib.expression.math.GS1Elements.VARIANT;

final class GS1Formatter
{
    private GS1Formatter() {
        throw new UnsupportedOperationException("util");
    }

    public static String format( String value, GS1Elements valueToReturn)
    {
        if( value == null )
            throw new IllegalArgumentException( "Can't extract data from null value" );

        if( value.length() < 3 )
            throw new IllegalArgumentException( "Value does not contains enough information" );

        String gs1Identifier = value.substring( 0, 3 );
        switch ( GS1Elements.fromKey( gs1Identifier ) ){
            case GS1_d2_IDENTIFIER:
            case GS1_Q3_IDENTIFIER:
                return formatValue(value, valueToReturn);
            case GS1_J1_IDENTIFIER:
            case GS1_d1_IDENTIFIER:
            case GS1_Q1_IDENTIFIER:
            case GS1_E0_IDENTIFIER:
            case GS1_E1_IDENTIFIER:
            case GS1_E2_IDENTIFIER:
            case GS1_E3_IDENTIFIER:
            case GS1_E4_IDENTIFIER:
            case GS1_I1_IDENTIFIER:
            case GS1_C1_IDENTIFIER:
            case GS1_e0_IDENTIFIER:
            case GS1_e1_IDENTIFIER:
            case GS1_e2_IDENTIFIER:
                throw new IllegalArgumentException( String.format( "gs1 identifier %s is not supported", gs1Identifier ) );
            default:
                throw new IllegalArgumentException( "Value does not start with a gs1 identifier" );
        }
    }

    private static String removeGS1Identifier( String value )
    {
        return value.substring( 3 );
    }

    private static String formatValue( String value, GS1Elements valueToReturn )
    {
        Map<String, String> dataMap = new HashMap<>();
        String[] gs1Groups = removeGS1Identifier( value ).split( GS1Elements.GS1_GROUP_SEPARATOR.getElement() );
        for ( String gs1Group : gs1Groups )
        {
            handleGroupData( gs1Group, dataMap );
        }
        if ( dataMap.containsKey( valueToReturn.getElement() ) )
        {
            return dataMap.get( valueToReturn.getElement() );
        }
        else
        {
            throw new IllegalArgumentException( "Required key does not exist for provided value" );
        }
    }

    private static void handleGroupData( String gs1Group, Map<String, String> dataMap )
    {
        if ( !gs1Group.isEmpty() )
        {
            int gs1GroupLength = gs1Group.length();
            String ai = GS1Elements.getApplicationIdentifier( gs1Group );
            Integer nextValueLength = AI_FIXED_LENGTH.get( ai.substring(0, 2) );
            if ( nextValueLength == null )
                nextValueLength = gs1GroupLength;
            dataMap.put( ai, gs1Group.substring( ai.length(), nextValueLength ) );
            handleGroupData( gs1Group.substring( nextValueLength ), dataMap );
        }
    }

    private static final Map<String, Integer> AI_FIXED_LENGTH = new HashMap<>();
    static {
        AI_FIXED_LENGTH.put(SSCC.getElement(), 20);
        AI_FIXED_LENGTH.put(GTIN.getElement(), 16);
        AI_FIXED_LENGTH.put(CONTENT.getElement(), 16);
        AI_FIXED_LENGTH.put("03", 16);
        AI_FIXED_LENGTH.put("04", 18);
        AI_FIXED_LENGTH.put(PROD_DATE.getElement(), 8);
        AI_FIXED_LENGTH.put(DUE_DATE.getElement(), 8);
        AI_FIXED_LENGTH.put(PACK_DATE.getElement(), 8);
        AI_FIXED_LENGTH.put("14", 8);
        AI_FIXED_LENGTH.put(BEST_BEFORE_DATE.getElement(), 8);
        AI_FIXED_LENGTH.put(SELL_BY.getElement(), 8);
        AI_FIXED_LENGTH.put(EXP_DATE.getElement(), 8);
        AI_FIXED_LENGTH.put("18", 8);
        AI_FIXED_LENGTH.put("19", 8);
        AI_FIXED_LENGTH.put(VARIANT.getElement(), 4);
        AI_FIXED_LENGTH.put("31", 10);
        AI_FIXED_LENGTH.put("32", 10);
        AI_FIXED_LENGTH.put("33", 10);
        AI_FIXED_LENGTH.put("34", 10);
        AI_FIXED_LENGTH.put("35", 10);
        AI_FIXED_LENGTH.put("36", 10);
        AI_FIXED_LENGTH.put("41", 16);
    }
}

File: src/commonMain/kotlin/org/hisp/dhis/lib/expression/math/GS1Formatter.kt
package org.hisp.dhis.lib.expression.math

import org.hisp.dhis.lib.expression.math.GS1Elements.Companion.fromKey
import org.hisp.dhis.lib.expression.math.GS1Elements.Companion.getApplicationIdentifier

object GS1Formatter {

    fun format(value: String?, valueToReturn: GS1Elements): String {
        requireNotNull(value) { "Can't extract data from null value" }
        require(value.length >= 3) { "Value does not contains enough information" }
        val gs1Identifier = value.substring(0, 3)
        return when (fromKey(gs1Identifier)) {
            GS1Elements.GS1_d2_IDENTIFIER, GS1Elements.GS1_Q3_IDENTIFIER -> formatValue(
                value,
                valueToReturn)

            GS1Elements.GS1_J1_IDENTIFIER, GS1Elements.GS1_d1_IDENTIFIER, GS1Elements.GS1_Q1_IDENTIFIER, GS1Elements.GS1_E0_IDENTIFIER, GS1Elements.GS1_E1_IDENTIFIER, GS1Elements.GS1_E2_IDENTIFIER, GS1Elements.GS1_E3_IDENTIFIER, GS1Elements.GS1_E4_IDENTIFIER, GS1Elements.GS1_I1_IDENTIFIER, GS1Elements.GS1_C1_IDENTIFIER, GS1Elements.GS1_e0_IDENTIFIER, GS1Elements.GS1_e1_IDENTIFIER, GS1Elements.GS1_e2_IDENTIFIER -> throw IllegalArgumentException(
                "gs1 identifier $gs1Identifier is not supported")

            else -> throw IllegalArgumentException("Value does not start with a gs1 identifier")
        }
    }

    private fun removeGS1Identifier(value: String): String {
        return value.substring(3)
    }

    private fun formatValue(value: String, valueToReturn: GS1Elements): String {
        val dataMap: MutableMap<String, String> = HashMap()
        val gs1Groups = removeGS1Identifier(value).split(GS1Elements.GS1_GROUP_SEPARATOR.element.toRegex())
            .dropLastWhile { it.isEmpty() }
            .toTypedArray()
        for (gs1Group in gs1Groups) {
            handleGroupData(gs1Group, dataMap)
        }
        return dataMap.getOrElse(
            valueToReturn.element
        ) { throw IllegalArgumentException("Required key does not exist for provided value") }
    }

    private fun handleGroupData(gs1Group: String, dataMap: MutableMap<String, String>) {
        if (gs1Group.isNotEmpty()) {
            val gs1GroupLength = gs1Group.length
            val ai = getApplicationIdentifier(gs1Group)
            var nextValueLength = AI_FIXED_LENGTH[ai.substring(0, 2)]
            if (nextValueLength == null) nextValueLength = gs1GroupLength
            dataMap[ai] = gs1Group.substring(ai.length, nextValueLength)
            handleGroupData(gs1Group.substring(nextValueLength), dataMap)
        }
    }

    private val AI_FIXED_LENGTH: MutableMap<String, Int> = HashMap()

    init {
        AI_FIXED_LENGTH[GS1Elements.SSCC.element] = 20
        AI_FIXED_LENGTH[GS1Elements.GTIN.element] = 16
        AI_FIXED_LENGTH[GS1Elements.CONTENT.element] = 16
        AI_FIXED_LENGTH["03"] = 16
        AI_FIXED_LENGTH["04"] = 18
        AI_FIXED_LENGTH[GS1Elements.PROD_DATE.element] = 8
        AI_FIXED_LENGTH[GS1Elements.DUE_DATE.element] = 8
        AI_FIXED_LENGTH[GS1Elements.PACK_DATE.element] = 8
        AI_FIXED_LENGTH["14"] = 8
        AI_FIXED_LENGTH[GS1Elements.BEST_BEFORE_DATE.element] = 8
        AI_FIXED_LENGTH[GS1Elements.SELL_BY.element] = 8
        AI_FIXED_LENGTH[GS1Elements.EXP_DATE.element] = 8
        AI_FIXED_LENGTH["18"] = 8
        AI_FIXED_LENGTH["19"] = 8
        AI_FIXED_LENGTH[GS1Elements.VARIANT.element] = 4
        AI_FIXED_LENGTH["31"] = 10
        AI_FIXED_LENGTH["32"] = 10
        AI_FIXED_LENGTH["33"] = 10
        AI_FIXED_LENGTH["34"] = 10
        AI_FIXED_LENGTH["35"] = 10
        AI_FIXED_LENGTH["36"] = 10
        AI_FIXED_LENGTH["41"] = 16
    }
}